roots_finding.py

```diff
@@ -0,0 +1,100 @@
+from typing import List
+from abc import ABC, abstractmethod
+from sympy.parsing.sympy_parser import parse_expr
+
+
+class LineSegment:
+    def __init__(self, left: float, right: float):
+        if right < left:
+            raise ValueError("The beginning of a line segment cannot exceed its end.")
+
+        self.left = left
+        self.right = right
+
+    @property
+    def length(self):
+        return self.right - self.left
+
+    @property
+    def center(self):
+        return (self.right + self.left) / 2
+
+    def __str__(self):
+        return f"[{self.left}, {self.right}]"
+
+    def __repr__(self):
+        return str(self)
+
+
+def separate_roots(function, line_segment: LineSegment,
+                   number_of_steps: int, variable: str = 'x') -> List[LineSegment]:
+    segments = []
+    step = line_segment.length / number_of_steps
+    cur_segment = line_segment
+    left_function_value = function.subs({variable: cur_segment.left})
+
+    for _ in range(number_of_steps):
+        cur_segment.right = cur_segment.left + step
+        right_function_value = function.subs({variable: cur_segment.right})
+        if left_function_value * right_function_value <= 0:  # todo: добавить погрешность?
+            segments.append(LineSegment(cur_segment.left, cur_segment.right))
+        left_function_value = right_function_value
+        cur_segment.left = cur_segment.right
+
+    return segments
+
+
+class SingleRootSolver(ABC):
+    @abstractmethod
+    def find_root(self, function, line_segment: LineSegment, accuracy, variable: str = 'x') -> float:
+        pass
+
+
+# It is assumed that there is exactly one non-multiple root inside the line segment
+class HalfDivisionSolver(SingleRootSolver):
+    def __init__(self):
+        self.number_of_iterations = 0
+
+    def find_root(self, function, line_segment: LineSegment, accuracy, variable: str = 'x') -> float:
```

Comment:
Please add typing for `accuracy`.

Reply:
No No

Reply:
🗿

Reply:
Okay, I thought and changed my mind. Will be done!